### ISSUE TITLE
Unreachable yield should still make a function a generator

### ISSUE THREAD

**Issue by kepler-5** [microsoft/pyright#5972]
**Describe the bug**
According to https://docs.python.org/3/reference/expressions.html#yieldexpr, when a function contains a `yield` expression, it becomes a generator function, regardless of the reachability of that `yield` expression. But in the code below, an unconditional `return` makes the `yield` unreachable and pyright seems to stop type checking as soon as it sees the `return`.

```python
def foo() -> None:
    return
    yield 42
```

This function signature is accepted by pyright but seems incorrect, since the presence of the `yield` keyword, even if unreachable, turns this function into a generator function. Calling `foo()` as written gives me `<generator object foo at 0x12dfd54e0>` rather than the promised `None`.

Removing the `return`, I get the expected error message:

```
Return type of generator function must be compatible with "Generator[Literal[42], Any, Any]"
  Type cannot be assigned to type "None"
```

I believe this to be a bug and I couldn't find a duplicate, but I'm definitely no expert so I'm sorry if I'm off about something here.

**VS Code extension or command-line**
Pylance v2023.9.10

**Comment by erictraut** [comment]
Pyright validates that the annotated return type of a generator function is compatible with `Generator` (or `AsyncGenerator`), but it currently does this only when type checkin a `yield` or `yield to` statement. In your example, the `yield` statement was determined to be unreachable, and errors detected when analyzing unreachable statements are suppressed. 

I've enhanced the logic that validates the annotated return address for a generator so it is performed regardless of whether a `yield` or `yield from` statement is evaluated. This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.328, which I just published. It will also be included in a future release of pylance.
